streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStreamThread.java

```diff
@@ -454,4 +480,19 @@ public void shutdown() {
     public Map<MetricName, Metric> consumerMetrics() {
         return Collections.unmodifiableMap(globalConsumer.metrics());
     }
+
+    public KafkaFuture<Uuid> globalConsumerInstanceId(final Duration timeout) {
```

Comment:
What are the threading rules here? Might we end up with the `clientInstanceId` and `fetchDeadline` being changed surprisingly by multiple calls to this method.

Reply:
The users thread would call `KafkaStreams#clientInstanceIds(...)` which is the only place calling this method. -- we could make `KafkaStreams#clientInstanceIds(...)` synchronized? In the end, `KafkaStreams` is supposed to be thread safe.

Reply:
Frankly, given that `fetchDeadline` might be modified (and pushed into the future) by a second call to `KafkaStreams#clientInstanceIds(...)` while the first was not completed yet, it seems we would need `synchronized` in addition?